### PR TITLE
Implement a helper to encode/decode human readable properties to compact properties #7

### PR DESCRIPTION
**Pull request by zyfrank**
…act properties #7

### CODE REVIEW

**Comment by zyfrank** [comment]
refact message fields mapping login

File: src/readable_compacts_transform_helper.js [hunk 1]
```diff
@@ -1,117 +1,273 @@
...
+    }
+    let remains = doCompactsReadableMapping(messageFieldsMapping, field['type'], pkgName, components);
+    if (remains !='')  {
+        readableName = readableName + '.' + remains;
+    }   
+    return readableName;
+};
+
+const strToComponents = (str) =>{
+    if (str.length > 128){
```
Comment (lucasvo) [inline src/readable_compacts_transform_helper.js]: This is not the correct behavior yet. If a string is used as a key in a map, we need to add our own protobuf extension to the field to specify the max length of the key and use that length, not 128 bytes.

File: src/readable_compacts_transform_helper.js [hunk 2]
```diff
@@ -1,117 +1,273 @@
...
+    let result ='';
+    for (let ii = 0; ii <16 ; ii++){
+        let component = coms.shift();
+        if (component.toInt() != 0){
+            result = result + transform(component);                       }
+    }
+    return result;
+};
+
+const doCompactsReadableMapping = (messageFieldsMapping, msgName, pkgName, compactComponents) => {
```
Comment (lucasvo) [inline src/readable_compacts_transform_helper.js]: In the go implementation we call the two different variations `literal` and `compact`. Could you replace "readable" with "literal" in the js code as well?

File: src/readable_compacts_transform_helper.js (hunk 1 above)
Comment (lucasvo) [reply src/readable_compacts_transform_helper.js]: Why is it 128 and not 64 bytes here?

File: src/readable_compacts_transform_helper.js [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+};
+
+const getField = (conditions, msgName, mappings) => {
+    let fields = mappings.get(msgName);
+    if (fields == undefined) {
+        throw new Error('No ' + msgName + ' definition:');
+    }
+    let field = fields.filter(itemFilter(conditions))[0];
+    if (field == undefined) {
+        throw new Error('Message ' + msgName + ' does not has corresponding field with conditions:' + JSON.stringify(conditions));
```
Comment (lucasvo) [inline src/readable_compacts_transform_helper.js]:
```suggestion
         throw new Error('Message ' + msgName + ' does not have corresponding field with conditions:' + JSON.stringify(conditions));
```

File: src/readable_compacts_transform_helper.js [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+        }
+    } else {
+        let msgs = jsonMeta;
+        for (let ii in msgs) {
+            results.set(outer_scope, msgs[ii]);
+        }
+    }
+};
+
+export class TransformHelper {
```
Comment (lucasvo) [inline src/readable_compacts_transform_helper.js]: Could you add [jsdoc](http://usejsdoc.org/about-getting-started.html) for the public method and add some comments here?

File: src/readable_compacts_transform_helper.js [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+    constructor(jsonMetaForProto, packageName, msgName) {
+        this.messageTypeNameToFields = new Map;
+        buildMessageNameToFieldsMapping(jsonMetaForProto, this.messageTypeNameToFields);
+        this.msgName = msgName;
+        this.pkgName = packageName;
+        if (this.messageTypeNameToFields.get(getQualifiedMsgName(msgName,packageName)) == undefined){
+            throw new Error(`No "Messsge:"  ${msgName} definition`);
+        }
+    }
+    compactsToReadableString(compactComponents) {
```
Comment (lucasvo) [inline src/readable_compacts_transform_helper.js]: What type are you using for compactProperty?

File: README.md [hunk 6]
```diff
@@ -112,6 +112,22 @@ const base64PreciseProofs = new PreciseProofs('base64');
...
+pbjs example.proto -t json -p "/usr/local/include" -o examples.json
+```
+To transform between Compacts format and Readable Name, you should provide generated json file and you should know the package name and the root message type defined in proto file.
+
+```
+import {TransformHelper} from "readable_compacts_transform_helper";
+var jsonFormat = require('examples.json');
+var helper = new TransformHelper(jsonFormat,'documents','ExampleDocument');
+helper.compactsToReadableString([3]);
+helper.readableStringToCompacts('value_bytes1')
```
Comment (lucasvo) [inline README.md]:
```suggestion
helper.literalPropertyToCompact('value_bytes1')
```

File: README.md [hunk 7]
```diff
@@ -112,6 +112,22 @@ const base64PreciseProofs = new PreciseProofs('base64');
...
+```
+pbjs example.proto -t json -p "/usr/local/include" -o examples.json
+```
+To transform between Compacts format and Readable Name, you should provide generated json file and you should know the package name and the root message type defined in proto file.
+
+```
+import {TransformHelper} from "readable_compacts_transform_helper";
+var jsonFormat = require('examples.json');
+var helper = new TransformHelper(jsonFormat,'documents','ExampleDocument');
+helper.compactsToReadableString([3]);
```
Comment (lucasvo) [inline README.md]:
```suggestion
helper.compactPropertyToLiteral([3]);
```

File: README.md [hunk 8]
```diff
@@ -112,6 +112,22 @@ const base64PreciseProofs = new PreciseProofs('base64');
...
+TransformHelper can be used to achieve this target
+First of all, you shoud install protobufjs to generate json definition from proto file.
+ 
+```
+pbjs example.proto -t json -p "/usr/local/include" -o examples.json
+```
+To transform between Compacts format and Readable Name, you should provide generated json file and you should know the package name and the root message type defined in proto file.
+
+```
+import {TransformHelper} from "readable_compacts_transform_helper";
```
Comment (lucasvo) [inline README.md]: Adding a few things on naming in the comments below to keep terminology in line with the go code base: 
Can you rename it to: PropertyTransformHelper?
The package `readable_compacts_transform_helper` seems a bit long. Why not make it part of the `precise-proofs` package?

File: src/readable_compacts_transform_helper.js [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+    } else {
+        let msgs = jsonMeta;
+        for (let ii in msgs) {
+            results.set(outer_scope, msgs[ii]);
+        }
+    }
+};
+
+export class TransformHelper {
+    constructor(jsonMetaForProto, packageName, msgName) {
```
Comment (lucasvo) [inline src/readable_compacts_transform_helper.js]: Trees can be constructed with a `ParentPrefix` which is prepended to every property. When converting from literal to compact this would have to be taken into account as well. For this we should add an option to the helper to specify a literal & compact prefix that is cut off from the property and replaced for each conversion.

See https://godoc.org/github.com/centrifuge/precise-proofs/proofs#TreeOptions for the usage in the backend.

**Review by lucasvo — COMMENTED** [review]
Looks good so far! Thanks for the work. I added some comments. The dynamic length keys still need some work as discussed in the issue on precise-proofs.

File: src/readable_compacts_transform_helper.js (hunk 1 above)
Comment (zyfrank) [reply src/readable_compacts_transform_helper.js]: here we have a few choices when golang side do encoding for map key (string or bytes):
1 padding the leading space so whole length is the max length of the key(protobuf extensions)
2 the leading compact component will be a length value (numbers of remaining components), with following components ( if key is string, one component(uint64) can contain 8 ascii char (0x**), if key is bytes, one components can contain 8 bytes), in this way it is impossible to detect leading 0x00 happened in bytes key
3 some subtle encoding  

in the js code, option 1 is used because the encoding rule is not finalized (I guess?)

File: src/readable_compacts_transform_helper.js (hunk 2 above)
Comment (zyfrank) [reply src/readable_compacts_transform_helper.js]: "ReadableName" is borrowed from go precise code,  if "literal" (better than "readable ") is the choice, it is fine

File: src/readable_compacts_transform_helper.js (hunk 4 above)
Comment (zyfrank) [reply src/readable_compacts_transform_helper.js]: will when most of the things is finalized

File: src/readable_compacts_transform_helper.js (hunk 1 above)
Comment (lucasvo) [reply src/readable_compacts_transform_helper.js]: The goal is to go with 1) and zero pad everything. I don't understand why leading zeroes couldn't be added though. The length of the compact key should always be fixed based on the protobuf option set on the field. E.g. a string with length 5 would be 5 bytes long (NB: bytes not characters, so unicode characters taking up more than one byte would so too in the key).

File: src/readable_compacts_transform_helper.js (hunk 9 above)
Comment (zyfrank) [reply src/readable_compacts_transform_helper.js]: logic and unittest for this added

File: src/readable_compacts_transform_helper.js (hunk 1 above)
Comment (zyfrank) [reply src/readable_compacts_transform_helper.js]: option "key_max_length" processing logic has been committed

Comment (lucasvo) [reply src/readable_compacts_transform_helper.js]: Ok, I understand what you're talking about now. In the literal notation we also always have to extend the value to the max length. I.e. fill it with 0x0 for every character that is missing. So technically a key that is not of the correct length is invalid and doesn't need to be handled. 

@andremarianiello would actually have to make sure this implemented on the go side.

Comment (zyfrank) [reply src/readable_compacts_transform_helper.js]: @andremarianiello 's  new commit has implemented this kind of logic
Acctually, I' m worry about map<**bytes**, >, if an original bytes key is started with byte 0x00,0x00,
JS cannot distinguish padding 0x00 0x00 0x00.......0x00  from original leading 0x00,0x00 because they are totoally mixed

Comment (lucasvo) [reply src/readable_compacts_transform_helper.js]: Even for bytes, the literal and compact bytes would always need to be zero padded. A 4 byte array would always be: 0x0, 0x0, 0x0a for example and never just 0x0a

Comment (zyfrank) [reply src/readable_compacts_transform_helper.js]: to simplify the discussion, assume following definition (**128** happenes at the same way 
map<bytes, string> bytes_map = 1 [
(propfs.max_key_length) = **16**
]

following 0x stand for binary, not hex string
there is an item which is [0x0,0x0,0x12,0x34]-> "blabla.."

if compacts mode on, when golang side do encoding for the proof of  this item, it will pad 12 of 0x0 for the key,   that is  0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,**0x0,0x0,0x12,0x34** (byte slice),then encode those 16 bytes to 2 uint64,  that is **[0x0,0x1234]** (attention, this is uint64 slice, not string slice) which is part of compact property(stand for the key [0x0,0x0,0x12,0x34]).

when JS part do transform from compacts to literal, as far as key is concerned, what it see is just **[0x0,0x1234]**.  It cannot determine the exact key bacause it just know the max length (16, 2 uint64), it does not know the actual length. 

string map does not suffer from this problem, because binary representation of a string will not begin with byte 0x00,0x00...  .    so JS part can remove lead padding 0x00...  safely

Comment (lucasvo) [reply src/readable_compacts_transform_helper.js]: I'm shortening your example to make it a little bit simpler:

```protobuf,
map<bytes, string> bytes_map = 1 [
(propfs.max_key_length) = 2
]
```

```go,
// NB this is go code, so the native type would be: []byte
var validLengthKey = []byte{0x00, 0x01}
// -> to literal:
var validLiteralKey = "0x0001"
```

The following is invalid and would be extended by the library when generating the tree:
```go,
var input = []byte{0x01}
// If such a value is fed to the go library, the key should be extended to the correct length
var literalKeyBytes = []byte{0x00, 0x01}
```

Also, we're using uint64 for protobuf field numbers (because that's the max value they can have) and for arrays. We don't need to make the mapping keys a multiple of 64 bits. For the compact type, I'd actually suggest we use https://github.com/MikeMcl/bignumber.js/ and not long. The goal is that the library returns a single array of bytes (most likely a bignumber is the best thing to return) not an array of uint64s.

Comment (andremarianiello) [reply src/readable_compacts_transform_helper.js]: @lucasvo Currently, all compact names are represented by some number of uint64s, which are then encoded into bytes. This made sense when all compact names were derived from uint64s, but now with map keys becoming part of the compact name, I think it might not make sense anymore.

Comment (zyfrank) [reply src/readable_compacts_transform_helper.js]: considering
map<bytes, string> bytes_map = 5 [
(propfs.max_key_length) = 2
] ,
now assume client of precise go library has two items for this map, and it want to produce 2 proofs for these two items
1.  [0x01] -> 'a'  
2.  [0x00,0x01] -> 'a'

After padding proccessing of go library, both items have same property defined in proof.proto 
for compact mode on, it is  array [0x5, 0x1, 0x61]   (here  0x5 is tag number, 0x61 is acsii of 'a' in hex)
for compact mode off, it is string "bytes_map[0x0001].a"
JS part cannot distinguish these two items as far as key ([0x1] or "0x0001") is concerned
actually, go library cannot distinguish these two proofs also. Although other fields in the proof are not same, but it do not know which item the proof is target for, it is a proof for [0x01] -> 'a'  ? or it is a proof for [0x00,0x01] -> 'a'?

Actually, we need another special processing just as **repeated** does (insert a length component in the compacts) 
for map<bytes,..>, consider following encoding rule,
when go library meet a **map**  item and it should produce components for the key in the proof,
it insert a length component (similar to length component of **repeated**),but this component contain two parts, the first 32 bits means numbers of following components which will contribute to this key, the second 32 bits means actual length (number of bytes) of the key.
Now for above two items, compact property will be:
1 [0x5, 0x1000000**01**,0x1,0x61]
2 [0x5, 0x1000000**02**,0x1,0x61]
literal property will be:
1 "bytes_map[0x01].a"
2 "bytes_map[0x0001].a"

in this way, we do not need "max_key_length" extension (of course if you want to limit the length of key, you can introduce this extension, i do not think it is needed)

As far as "Long" is used instead of "bignumber", JS part need parse(here we use protobufjs library) the binary generatey by proto buffer marshal for **proof** message ,  when it meet a uint64(that is the type of compact component) , it will decoded it as Long
 @lucasvo 
@andremarianiello

Comment (zyfrank) [reply src/readable_compacts_transform_helper.js]: even we change type of compact component from uint64 to byte, we cannot cope with problem mentioned in above comments when map<bytes,..> is met

Comment (zyfrank) [reply src/readable_compacts_transform_helper.js]: consider encoding of length of  **repeated** field,
if we use byte compact instead of uint64 compact
we have to introduce length of length concept, because length of **repeated** can be 0xaf,  0x01af and 0x0100af

Comment (lucasvo) [reply src/readable_compacts_transform_helper.js]: I'm sorry if my earlier comment was not clear. The key will always have the specified length. When creating the compact properties, if the precise-proofs library encounters a shorter key it would extend it to the length specified. To take your examples:

```protobuf,
map<bytes, string> bytes_map = 5 [
(propfs.max_key_length) = 2
];
```
> now assume client of precise go library has two items for this map, and it want to produce 2 proofs for these two items
> 
> [0x01] -> 'a'
> [0x00,0x01] -> 'a'

The library here would complain about the first item ([0x01]) being invalid because it has an invalid length and not actually allow you to generate the tree. You wouldn't end up with a tree that has two leaves:
1) `[0x05, 0x01]`
2) `[0x05, 0x00, 0x01]`

Does that make sense?

**Comment by lucasvo** [comment]
> As far as "Long" is used instead of "bignumber", JS part need parse(here we use protobufjs library) the binary generatey by proto buffer marshal for proof message , when it meet a uint64(that is the type of compact component) , it will decoded it as Long

It took me some time to figure out what you were referring too and then I realized that you were referring to this:

```protobuf,
message Proof {
  oneof property {
    string readable_name = 7;
    FieldNums compact_name = 9;
  }
```

This was chosen initially when we created the new format without support for mappings. At that point it was simple, each field would be a simple uint64. However with mappings, that is not the case anymore so we'll change the proof format to:

```
message Proof {
  oneof property {
    string literal_name = 7;
    bytes  compact_name = 9;
  }
```

So the literal name doesn't need to be a multiple of 64 bytes anymore. In Javascript we can just accept a single hex string as the compact name and convert it internally to bytes or bignumber and then start parsing it byte by byte

**Comment by zyfrank** [comment]
if length of key is fixed (for map <bytes, ..>), then it's fine for go precise proof library and js, here go precise library does not have concept of padding, client of the library should pad (or all map item key's length is fixed)because go library expect fixed length. extension "max_key_length" should be renamed to "key_length" to avoid misunderstanding.

________________________________
发件人: Lucas Vogelsang <notifications@github.com>
发送时间: 2018年12月31日 8:50
收件人: centrifuge/precise-proofs-js
抄送: zyfrank; Author
主题: Re: [centrifuge/precise-proofs-js] WIP: Implement a helper to encode/decode human readable properties to compact properties #7 (#8)


@lucasvo commented on this pull request.

________________________________

In src/readable_compacts_transform_helper.js<https://github.com/centrifuge/precise-proofs-js/pull/8#discussion_r244563898>:

> +            tmp = first16ComsToString(components,(com)=>{ return hextoString(com.toString(16));});
+            readableName = readableName + '[' + tmp + ']';
+            break;
+        default:
+            throw new Error(`Invalid key type: ${field['keytype']}`);
+        }
+    }
+    let remains = doCompactsReadableMapping(messageFieldsMapping, field['type'], pkgName, components);
+    if (remains !='')  {
+        readableName = readableName + '.' + remains;
+    }
+    return readableName;
+};
+
+const strToComponents = (str) =>{
+    if (str.length > 128){


I'm sorry if my earlier comment was not clear. The key will always have the specified length. When creating the compact properties, if the precise-proofs library encounters a shorter key it would extend it to the length specified. To take your examples:

map<bytes, string> bytes_map = 5 [
(propfs.max_key_length) = 2
];

now assume client of precise go library has two items for this map, and it want to produce 2 proofs for these two items

[0x01] -> 'a'
[0x00,0x01] -> 'a'

The library here would complain about the first item ([0x01]) being invalid because it has an invalid length and not actually allow you to generate the tree. You wouldn't end up with a tree that has two leaves:

  1.  [0x05, 0x01]
  2.  [0x05, 0x00, 0x01]

Does that make sense?

―

**Comment by zyfrank** [comment]
if type of compact_name is changed from FieldNum(uint64 slice ) to bytes, consider compact property encoding for following def.

map<uint32,..>
map<sint32,...>
map<uint64,..>
map<sint64,...>
... for other integer
map<string,...>
repeated<...>

it will introduce length of length concept.

________________________________
发件人: Lucas Vogelsang <notifications@github.com>
发送时间: 2018年12月31日 8:59
收件人: centrifuge/precise-proofs-js
抄送: zyfrank; Author
主题: Re: [centrifuge/precise-proofs-js] WIP: Implement a helper to encode/decode human readable properties to compact properties #7 (#8)


As far as "Long" is used instead of "bignumber",  part need parse(here we use protobufjs library) the binary generatey by proto buffer marshal for proof message , when it meet a uint64(that is the type of compact component) , it will decoded it as Long

It took me some time to figure out what you were referring too and then I realized that you were referring to this:

message Proof {
  oneof property {
    string readable_name = 7;
    FieldNums compact_name = 9;
  }

This was chosen initially when we created the new format without support for mappings. At that point it was simple, each field would be a simple uint64. However with mappings, that is not the case anymore so we'll change the proof format to:

message Proof {
  oneof property {
    string literal_name = 7;
    bytes  compact_name = 9;
  }


So the literal name doesn't need to be a multiple of 64 bytes anymore. In Javascript we can just accept a single hex string as the compact name and convert it internally to bytes or bignumber and then start parsing it byte by byte

―

**Comment by lucasvo** [comment]
> if type of compact_name is changed from FieldNum(uint64 slice ) to bytes, consider compact property encoding for following def. 
> map<uint32,..>
> map<sint32,...>
> map<uint64,..>
> map<sint64,...>
> ... for other integer
> map<string,...>
> repeated<...>
> it will introduce length of length concept.

Why would it introduce the length of length concept? uint32, uint64 etc have a fixed length already and with our extension we'd give strings & bytes a fixed length as well. Could you maybe give a clearer example? I'm not sure I know what you mean.

**Comment by lucasvo** [comment]
@zyfrank while it would be nice to compress it by prefixing it with the length, the additional complexity of it is not desired. The fixed length is important so bytes can be accessed directly without having to parse the byte stream to check the length of previous parts of the property. E.g:

```
// Variable length:
L_A= length marker
(L_A=4)AAAABBBB => 4AAAABBBB
(L_A=2)AABBBB => 2AABBBB

// Fixed length:
AAAABBBB
00AABBBB
```

This would break an important aspect of the structure: that to get BBBB you only need to read position [4:7] to get BBBB.

**Comment by zyfrank** [comment]
It's.  Just to make sure uint64[] compact will be replaced with byte[] compact so I can change the corresponding part in JS side

________________________________
发件人: Lucas Vogelsang <notifications@github.com>
发送时间: 2019年1月2日 17:25
收件人: centrifuge/precise-proofs-js
抄送: zyfrank; Mention
主题: Re: [centrifuge/precise-proofs-js] WIP: Implement a helper to encode/decode human readable properties to compact properties #7 (#8)


@zyfrank<https://github.com/zyfrank> while it would be nice to compress it by prefixing it with the length, the additional complexity of it is not desired. The fixed length is important so bytes can be accessed directly without having to parse the byte stream to check the length of previous parts of the property. E.g:

// Variable length:
L_A= length marker
(L_A=4)AAAABBBB => 4AAAABBBB
(L_A=2)AABBBB => 2AABBBB

// Fixed length:
AAAABBBB
00AABBBB


This would break an important aspect of the structure: that to get BBBB you only need to read position [4:7] to get BBBB.

―

**Comment by lucasvo** [comment]
@zyfrank yes, we will use []byte for the compact form.

**Comment by zyfrank** [comment]
@lucasvo @andremarianiello 
How about proto buffer message tag number  encoding? it will occupy fixed 2 byes or fixed 4 bytes?

**Comment by lucasvo** [comment]
Protobuf message tag numbers are no larger than 2**29-1. Rounding it up to the next 8bit multiple we'd get to 32 bits or 4 bytes. That's the length of a message tag.

**Review by lucasvo — APPROVED** [review]
A few smaller comments. This implements everything discussed. Though we'll need to refactor it a little bit once we change from uint64 to just arbitrary lenght byte slices. I'm ok merging this and doing the other changes later.

**Comment by lucasvo** [comment]
@zyfrank Andre has renamed the `proofs.key_max_length` extension to `proofs.key_length`. Could you update this PR accordingly before I merge it?

**Comment by zyfrank** [comment]
sure

File: test/PropertyTransformHelper.test.js [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+            (new PropertyTransformHelper(jsonMetaFormat, packageName, msgName,literalPrefix,compactPrefix)).literalToCompactProperty(literalWithErrorPrefix);
+        }).toThrow();
+    })
+
+    it('string map should work', () => {
+        expect((new PropertyTransformHelper(jsonMetaFormat, packageName, msgName)).compactPropertyToLiteral(compactPropertyForStringMap)).toEqual(literalForStringMap);
+        expect((new PropertyTransformHelper(jsonMetaFormat, packageName, msgName)).compactPropertyToLiteral(compactPropertyForStringMap2)).toEqual(literalForStringMap2);
+        expect((new PropertyTransformHelper(jsonMetaFormat, packageName, msgName)).literalToCompactProperty(literalForStringMap)).toEqual(compactPropertyForStringMap);
+    });
+
```
[the remark this answers is not in the document]
Comment (zyfrank) [reply test/PropertyTransformHelper.test.js]: yes, it should be

File: README.md [hunk 11]
```diff
@@ -112,6 +112,22 @@ const base64PreciseProofs = new PreciseProofs('base64');
...
+PropertyTransformHelper can be used to achieve this target
+First of all, you shoud install protobufjs to generate json definition from proto file.
+ 
+```
+pbjs example.proto -t json -p "/usr/local/include" -o examples.json
+```
+To transform between Compact Property and Literal String, you should provide generated json file and you should provide the package name and the message type defined in proto file.
+
+```
+import {PropertyTransformHelper} from "PropertyTransformHelper";
```
[the remark this answers is not in the document]
Comment (zyfrank) [reply README.md]: right, should be imported `from precise-proofs`

[the remark this answers is not in the document]
Comment (zyfrank) [reply README.md]: about “ uint64 to just arbitrary lenght byte slices.” I'll try to do corresponding change when go lang side is ready

**Comment by lucasvo** [comment]
@zyfrank, I'll review your changes soon. @andremarianiello has updated the code in this PR: https://github.com/centrifuge/precise-proofs/pull/39/files I've taken a first look and it should be pretty much the final implementation. So if you'd like, you could use that already. 

When working with bytes, I think the cleanest interface would be to use hex encoded strings (0xabcdef012345679). Could you adjust the transform helper to take in strings instead of arrays of ints?

As you might have noticed on the bounty, I've increased the bounty by 200 DAI to make up for the back and forth and the probably bigger than anticipated scope. Thanks a lot for your contribution so far. When you're done with this bounty, I'd love to point you to our other bounties that you might like to do (https://gitcoin.co/issue/centrifuge/precise-proofs/38/2050 & https://gitcoin.co/issue/centrifuge/precise-proofs/34/2049)

**Review by lucasvo — APPROVED** [review]
Approved. Last step would be to add the last changes implemented on the go side.

**Comment by lucasvo** [comment]
Just replied to your comments on the other PR, @zyfrank and merged that PR. Is there anything else you need to know to implement the change to bytes?

File: src/PropertyTransformHelper.js [hunk 12]
```diff
@@ -63,26 +63,32 @@ const doCompactPropertyLiteralMapping = (messageFieldsMapping, msgName, pkgName,
     if (compacts.remaining() == 0){
         return '';
     }
-    let tagNumber = compacts.readUint32();
```
Comment (lucasvo) [inline src/PropertyTransformHelper.js]: Why did you change to uint64? Protobuf tags should be only 32 bytes. Is the go implementation faulty?

Comment (lucasvo) [reply src/PropertyTransformHelper.js]: I just realized that it indeed is. This is due to us using the same type for slice indexes and tag names. We should fix that in Go and only use uint64 for slice length. Would you like to fix that if I increase the bounty on this a bit more? It would be great to get this merged in at the same time. If not though it's ok to commit it like that.

Comment (zyfrank) [reply src/PropertyTransformHelper.js]: @lucasvo change is done,  actually I also make change for GO side on my forked repo, just do not know vedhavyas have done the work.

Comment (zyfrank) [reply src/PropertyTransformHelper.js]: vedhavyas's PR is for another problem, have submitted a PR on Go side

**Review by lucasvo — APPROVED** [review]
The title of the PR is still WiP but this looks largely done. Do you think you'll be doing anymore changes?

**Comment by zyfrank** [comment]
@lucasvo  currrently js part is at the same page with go side (added enum surport need no change in js part), 
so no more change is needed currently, I'll keep eyes on those two repos.
